Add documentation for QuickStart.java (#7730)
cvc5/cvc5: docs/api/java/quickstart.rst

Quickstart Guide
================

> First, create a cvc5 `Solver <io/github/cvc5/api/Solver.html>`_ instance using try with resources:

Reviewer: Nit: wrap line

Author: Done

.. code-block:: java

     try (Solver solver = new Solver())
     {
       /** write your code here */
     }

> We will ask the solver to produce models and unsat cores in the following,
> and for this we have to enable the following options.

Reviewer: ```suggestion
To produce models and unsat cores, we have to enable the following options.
```

(I know that this is just copied from the other guides, but the wording is a bit verbose)

.. literalinclude:: ../../../examples/api/java/QuickStart.java
     :language: java
     :lines: 32-33

Next we set the logic.

> The simplest way to set a logic for the solver is to choose "ALL".

Reviewer: ```suggestion
The simplest way to set a logic for the solver is to choose ``"ALL"``.
```

This enables all logics in the solver.
Alternatively, ``"QF_ALL"`` enables all logics without quantifiers.
To optimize the solver's behavior for a more specific logic,
use the logic name, e.g. ``"QF_BV"`` or ``"QF_AUFBV"``.

.. literalinclude:: ../../../examples/api/java/QuickStart.java
     :language: java
     :lines: 42

> In the following, we will define constraints of reals and integers.

Reviewer: ```suggestion
In the following, we will define real and integer constraints.
```

For this, we first query the solver for the corresponding sorts.

.. literalinclude:: ../../../examples/api/java/QuickStart.java
     :language: java
     :lines: 46-47

Now, we create two constants ``x`` and ``y`` of sort ``Real``,
and two constants ``a`` and ``b`` of sort ``Integer``.

> Notice that these are *symbolic* constants, but not actual values.

Reviewer: ```suggestion
Notice that these are *symbolic* constants, not actual values.
```

.. literalinclude:: ../../../examples/api/java/QuickStart.java
     :language: java
     :lines: 101-102

It is also possible to get values for terms that do not appear in the original
formula.

.. literalinclude:: ../../../examples/api/java/QuickStart.java
     :language: java
     :lines: 106-107

We can convert these values to Java types.

.. literalinclude:: ../../../examples/api/java/QuickStart.java
     :language: java
     :lines: 109-116

Another way to independently compute the value of ``x - y`` would be to
perform the (rational) arithmetic manually.
However, for more complex terms, it is easier to let the solver do the
evaluation.

.. literalinclude:: ../../../examples/api/java/QuickStart.java
     :language: java
     :lines: 122-134

This will print:

.. code:: text

  computed correctly

Next, we will check satisfiability of the same formula,
only this time over integer variables ``a`` and ``b``.
For this, we first reset the assertions added to the solver.

.. literalinclude:: ../../../examples/api/java/QuickStart.java
     :language: java
     :lines: 140

Next, we assert the same assertions as above, but with integers.
This time, we inline the construction of terms

> to the assertion command.

Reviewer: ```suggestion
in the assertion command.
```

in the assertion command.

.. literalinclude:: ../../../examples/api/java/QuickStart.java
     :language: java
     :lines: 145-149
     :dedent: 6

Now, we check whether the revised assertion is satisfiable.

.. literalinclude:: ../../../examples/api/java/QuickStart.java
     :language: java
     :lines: 152-156
     :dedent: 6

This time the asserted formula is unsatisfiable:

.. code:: text

  expected: unsat
  result: unsat

We can query the solver for an unsatisfiable core, that is, a subset
of the assertions that is already unsatisfiable.

.. literalinclude:: ../../../examples/api/java/QuickStart.java
     :language: java
     :lines: 160-166
     :dedent: 6

This will print:

.. code:: text

  unsat core size: 3
  unsat core:
  (< 0 a)
  (< 0 b)
  (< (+ a b) 1)

Example
-------

| The SMT-LIB input for this example can be found at `examples/api/smtlib/quickstart.smt2 <https://github.com/cvc5/cvc5/blob/master/examples/api/smtlib/quickstart.smt2>`_.
| The source code for this example can be found at `examples/api/java/QuickStart.java <https://github.com/cvc5/cvc5/blob/master/examples/api/java/QuickStart.java>`_.

.. api-examples::

>     <examples>/api/cpp/quickstart.cpp
>     <examples>/api/java/QuickStart.java

Reviewer 2: Show the java code by default.
```suggestion
    <examples>/api/java/QuickStart.java
    <examples>/api/cpp/quickstart.cpp
```
